enterprise/server/backends/migration_cache/migration_cache.go

```diff
@@ -0,0 +1,157 @@
+	migrateAndExit            = flag.Bool("cache.migration_cache.migrate_and_exit", false, "If true, attempt to migrate src cache to dest cache and exit upon completion.")
+	clearCacheBeforeMigration = flag.Bool("cache.migration.clear_cache_before_migration", false, "If set, clear any existing cache content in the dest cache.")
+	autoFixDoubleReadErrs     = flag.Bool("cache.migration.auto_fix_double_read_errs", false, "If set, automatically fix any double read errs in the dest cache")
+)
+
+type MigrationCache struct {
+	Src        interfaces.Cache
+	Dest       MigrationDestCache
+	SrcRootDir string
+
+	mu             sync.RWMutex
+	eg             *errgroup.Group
+	doubleReadChan chan *repb.Digest
+}
+
+type MigrationDestCache interface {
+	interfaces.Cache
+
+	// Implementation should be async, so that the double write doesn't slow down operations on the primary db
+	// with 2 synchronous writes
+	DoubleWrite(ctx context.Context, d DoubleWriteData) error
+}
+
+type DoubleWriteData struct {
+	digest *repb.Digest
+	add    *Add
+	remove *Remove
+}
+
+type Add struct {
+	Data []byte
+}
+
+type Remove struct{}
+
+func Register(env environment.Env) error {
+	mc := env.GetMigrationCache()
+	if mc.Src == nil && mc.Dest == nil {
+		return nil
+	} else if mc.Src == nil || mc.Dest == nil {
+		// Return error - both must be set
+	}
+
+	ctx := context.Background()
+	mc.eg = &errgroup.Group{}
+	env.SetCache(mc)
+
+	if *migrateAndExit {
+		if err := mc.Migrate(ctx, mc.srcRootDir); err != nil {
+			log.Errorf("Migration failed: %s", err)
+			os.Exit(1)
+		}
+		os.Exit(0)
+	}
+
+	mc.eg.Go(func() error {
+		mc.ProcessDoubleReads(context.Background())
+		return nil
+	})
+
+	return nil
+}
+
+func (c *MigrationCache) Get(ctx context.Context, d *repb.Digest) ([]byte, error) {
+	shouldDoubleRead := rand.Float64() <= *doubleReadPercentage
+	if shouldDoubleRead {
+		c.doubleReadChan <- d
+	}
+
+	return c.Src.Get(ctx, d)
+}
+
+func (c *MigrationCache) Set(ctx context.Context, d *repb.Digest, data []byte) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if err := c.Src.Set(ctx, d, data); err != nil {
+		return err
+	}
+	if err := c.Dest.DoubleWrite(ctx, DoubleWriteData{
+		digest: d,
+		add:    &Add{Data: data},
+	}); err != nil {
+		// Log error but don't fail the call
+	}
+	return nil
+}
+
+func (c *MigrationCache) ProcessDoubleReads(ctx context.Context) {
+	for digest := range c.doubleReadChan {
+		c.eg.Go(func() error {
+			c.mu.Lock()
+			defer c.mu.Unlock()
+
+			primaryData, err := c.Src.Get(ctx, digest)
+			secondaryData, err := c.Dest.Get(ctx, digest)
+
+			if primaryData != secondaryData {
+				// Log so we can identify and debug double read errors
+			}
+
+			//
+			if *autoFixDoubleReadErrs {
+				c.Dest.Set(ctx, digest, primaryData)
+			}
+		})
+	}
+}
+
+// Migrate copies all data from the src cache to the dest cache
+func (c *MigrationCache) Migrate(ctx context.Context, srcRootDir string) error {
```

Comment:
One issue we've seen in the past is that Migrating a single app's worth of data can take quite a long time. With 16 apps I think we predicted a full serial migration, app by app, would take 2+ days. During that time a single app would be down.

What do you think about having an "online" migration mode where files are copied between caches in the background while the app runs? This would allow all of the caches to migrate simultaneously while serving traffic.

Reply:
Side note that doesn't affect your point - can you clarify what you mean by "app" here? 

Yeah that's a good idea. And we could make sure to buffer the online migration so it doesn't monopolize resources or the lock while we're trying to serve traffic

Reply:
We can apply different configs to each server, right? filepath.WalkDir walks in lexical order, so it really only makes sense to have 1 server copying data at a time

Reply:
By "app" I mean one cache server.

I agree with your point about ensuring the online migration does not hog resources. We've seen this be an issue in the past, so having a builtin way to ratelimit it will be important.

Will ping you re filepath issue, but I think it's not relevant here because each server stores different data.